optima/optimization.py

```diff
@@ -166,16 +269,39 @@ def minoutcomes(project=None, optim=None, inds=0, maxiters=1000, maxtime=None, v
     if max(inds)>lenparlist: raise OptimaException('Index %i exceeds length of parameter list (%i)' % (max(inds), lenparlist+1))
     tvec = project.settings.maketvec(end=objectives['end']) # WARNING, this could be done better most likely
     
+    # Handle budget and remove fixed costs
     totalbudget = objectives['budget']
-    nprogs = len(progset.programs)
+    progkeys = progset.programs.keys()
+    optiminds = findinds(progset.optimizable())
+    fixedinds = findinds(1-array(progset.optimizable()))
+    nprogs = len(optiminds) # Only count optimizable programs
     budgetvec = progset.getdefaultbudget()[:]
+    
+    # Error checking
     if isnan(budgetvec).any():
-        budgetlessprograms = array(progset.programs.keys())[isnan(budgetvec)].tolist()
-        output = 'WARNING!!!!!!!!!! Not all programs have a budget associated with them.\n A uniform budget will be used instead.\n Programs with no budget are:\n'
-        output += '\n'.join(budgetlessprograms)
-        print(output)
-        budgetvec = zeros(nprogs)+totalbudget/nprogs
-    else: budgetvec *= totalbudget/sum(budgetvec) # Rescale
+        errormsg = 'Program "%s" does not have any budget' % progkeys[findinds(isnan(budgetvec))]
```

Comment:
this would raise an error in cases where people had not specified a budget for a fixed-cost program, which may not be strictly speaking a problem for the subsequent steps.